examples/common/convert_c4.py

```diff
@@ -4,105 +4,359 @@
+                        if n_samples == self.expected_num_samples:
+                            return
+
+
+class TokenizedC4(ShardedC4):
+    """A tokenized, concatenated, iterable dataset.
+
+    To use data created by this class and written to MDS format:
+
+    ```python
+        import torch
+        from streaming.base import StreamingDataset
+        from transformers import AutoTokenizer
+
+        tokenizer = AutoTokenizer.from_pretrained('your/tokenizer')
+        ds = StreamingDataset(local='mds-data-folder', split='val')
+
+        # note, you need to copy the numpy array because the original is non-writeable
+        # and torch does not support non-writeable tensors, so you get a scary warning and
+        # if you do try to write to the tensor you get undefined behavior
+        tokens = torch.from_numpy(np.frombuffer(ds[0]['tokens'], dtype=np.int64).copy())
+        print(tokenizer.decode(tokens))
+    ```
+    """
+
+    def __init__(self, tokenizer: PreTrainedTokenizerBase, *args, **kwargs):
+        self.tokenizer = tokenizer
+        super().__init__(*args, **kwargs)
+
+    def generate_samples(self) -> Iterable[Dict[str, bytes]]:
+        """Generator over each dataset sample.
+
+        Args:
+            samples (IterableDataset): An iterable dataset that is multi-worker compatible
+
+        Yields:
+            Sample dicts.
+        """
+
+        n_samples = 0
+        buffer = {}
+        while True:
+            for batch in self.loader:
+                encoded = self.tokenizer(batch['text'], truncation=False, padding=False)
+                for iids in encoded['input_ids']:
+                    buffer['tokens'] = buffer.get('tokens', []) + iids
+                    while len(buffer['tokens']) >= self.max_length:
+                        concat_sample = {}
+                        concat_sample['tokens'] = buffer['tokens'][:self.
+                                                                max_length]
+                        buffer['tokens'] = buffer['tokens'][self.max_length:]
+
+                        yield {
+                            # convert to bytes to store in MDS binary format
+                            'tokens': np.asarray(concat_sample['tokens']).tobytes()
+                        }
+                        n_samples += 1
+                        if n_samples == self.expected_num_samples:
+                            return
+
+
+def build_hf_c4_dataset(
+        split: str,
+        mode: ConcatMode,
+        max_length: Optional[int] = None,
+        sep_text: Optional[str] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        expected_num_samples: Optional[int] = None) -> IterableDataset:
     """Collect the samples for this dataset split.
 
     Args:
         split (str): Split name.
+        mode (ConcatMode): NO_CONCAT, CONCAT_TEXT, or CONCAT_TOKENS
+        sep_text (str): if mode is CONCAT_TEXT, what string to use to separate samples
```

Comment:
left comment above, but maybe here is a better place. should `CONCAT_TOKENS` also support `sep_text`?

Reply:
I was imagining relying on the tokenizer to add special tokens, but this may be around the way GPT2's tokenizer does not do that. Is that what you are thinking?

Reply:
so you assume that the tokenizer adds special tokens including EOS? we should probably at least have a way to not require that? but i haven't thought about this as much as you, so trust what you think.

Reply:
the default behavior of huggingface tokenizers is to pass `add_special_tokens=True`... but we know that e.g. GPT2 does not add EOS or BOS. If the tokenizer you pass is like GPT2 and it doesn't add either, currently I raise an exception, so we can be sure that we won't pre-tokenize concat without separators. But to support GPT2, we'd need to do what you suggest and if they pass `--sep_text="<|endoftext|>"`, add that.

we should also let them pass `--sep_text_position`, right now I assume they want BOS

Reply:
all good points. ok, as long as you error and not silently fail. seems like something we can fix after the bones of this are in.

Reply:
although maybe we really need to support the gpt2 tokenizer? not sure

Reply:
I changed sep_text to bos_text and eos_text and didn't make it exclusive with tokenization, so now we can handle GPT2 if the user gives us `--bos_text=<|endoftext|>`